--- venv/lib/python3.13/site-packages/pittgoogle/pubsub.py	2026-03-18 06:53:09.896979079 -0700
+++ pubsub.py	2026-03-18 07:14:03.856635670 -0700
@@ -1,4 +1,4 @@
-# -*- coding: UTF-8 -*-
+# -*- coding: utf-8 -*-
 """Classes to facilitate connections to Google Cloud Pub/Sub streams.

 .. autosummary::
@@ -13,6 +13,7 @@
 import concurrent.futures
 import datetime
 import logging
+import multiprocessing.connection
 import queue
 import time
 from typing import Any, Callable, List, Literal, Optional, Union
@@ -578,7 +579,12 @@
             self._executor = concurrent.futures.ThreadPoolExecutor(self.max_workers)
         return self._executor

-    def stream(self, block: bool = True) -> None:
+    def stream(self, block: bool = True,
+               pipe: multiprocessing.connection.Connection = None,
+               heartbeat: int = 60,
+               max_runtime: datetime.timedelta = None,
+               max_nmsgs: int = None
+               ) -> int:
         """Open the stream in a background thread and process messages through the callbacks.

         Recommended for long-running listeners.
@@ -588,19 +594,36 @@
                 Whether to block the main thread while the stream is open. If `True`, block
                 indefinitely (use `Ctrl-C` to close the stream and unblock). If `False`, open the
                 stream and then return (use :meth:`~Consumer.stop()` to close the stream).
-                This must be `True` in order to use a `batch_callback`.
+                This must be `True` in order to use a `batch_callback`, and if you use
+                any of pipe, max_runtime, or max_nmsgs.
+            pipe : `multiprocessing.connection.Connection`
+                A pipe.  Will listen for a dictionary on this pipe; if { 'command': 'die' }
+                is received will exit.  Will also send a heartbeat message with
+                { 'message': 'ok', 'nconsumed': int, 'runtime': datetime.timedelta } every
+                heartbeat seconds.
+            heartbeat : Time out after this many seconds have elapsed; check the pipe for
+                a command, and send the heartbeat message
+            max_runtime : datetime.timedelta, default None
+                If not None, stop after running this long.
+            max_nmsgs : int, default None
+                If not None, stop after receiving this many messages.
+
         """
+        if ( not block ) and any( i is not None for i in [ pipe, max_runtime, max_nmsgs ] ):
+            raise RuntimeError( "Use of any of pipe, max_runtime, or max_nmsgs requires block" )
+
         # open a streaming-pull and process messages through the callback, in the background
         self._open_stream()

         if not block:
-            msg = "The stream is open in the background. Use consumer.stop() to close it."
+            msg = "The stream is open in the background. Use consumer.stop() to close it prematurely"
             print(msg)
             LOGGER.info(msg)
-            return
+            return None

         try:
-            self._process_batches()
+            return self._process_batches( pipe=pipe, heartbeat=heartbeat,
+                                          max_runtime=max_runtime, max_nmsgs=max_nmsgs )

         # catch all exceptions and attempt to close the stream before raising
         except (KeyboardInterrupt, Exception):
@@ -634,42 +657,115 @@
         else:
             message.nack()

-    def _process_batches(self):
+    def _drain_queue( self, batch: list = [] ) -> int:
+        while not self._queue.empty():
+            batch.append( self._queue.get( block=True ) )
+        self.batch_callback( batch )
+        return len( batch )
+
+
+    def _process_batches(self,
+                         pipe: multiprocessing.connection.Connection = None,
+                         heartbeat: int = 60,
+                         max_runtime: datetime.timedelta = None,
+                         max_nmsgs: int = None
+                         ) -> int:
         """Run the batch callback if provided, otherwise just sleep.

-        This never returns -- it runs until it encounters an error.
+        Runs until max_runtime has elapsed or until max_nmsgs have been
+        processed, whichever comes first.  If both are None, runs
+        indefinitely.
+
         """
+
         # if there's no batch_callback there's nothing to do except wait until the process is killed
         if self.batch_callback is None:
+            # LOGGER.warning( "There's no batch_callback, so _process_batches isn't doing anything!" )
+            raise RuntimeError( "There's no batch_callback, so _process_batches won't do anything!" )
+            t0 = datetime.datetime.now()
             while True:
-                time.sleep(60)
+                time.sleep( heartbeat )
+                if pipe is not None:
+                    if pipe.poll():
+                        msg = pipe.recv()
+                        if ( 'command' in msg ) and ( msg['command'] == 'die' ):
+                            self.stop()
+                            return 0
+                    pipe.send( { "message": "ok", "nconsumed": 0, "runtime": datetime.datetime.now() - t0 } )
+            return

         batch, count = [], 0
+        totprocessed = 0
+        tstart = time.monotonic()
+        t0 = tstart
+        max_runtime = max_runtime.total_seconds() if max_runtime is not None else None
+        firstheartbeat = datetime.datetime.now()
+        lastheartbeat = firstheartbeat
         while True:
+            t1 = time.monotonic()
+            waittime = max( min( self.batch_max_wait_between_messages, heartbeat ) - (t1-t0), 0 )
             try:
                 batch.append(
-                    self._queue.get(block=True, timeout=self.batch_max_wait_between_messages)
+                    self._queue.get(block=True, timeout=waittime)
                 )

             except queue.Empty:
                 # hit the max wait. process the batch
-                self.batch_callback(batch)
+                t0 = t1
+                if len(batch) > 0:
+                    self.batch_callback(batch)
+                totprocessed += len(batch)
                 batch, count = [], 0

             # catch anything else and try to process the batch before raising
             except (KeyboardInterrupt, Exception):
-                self.batch_callback(batch)
+                LOGGER.error( f"Got some kind of exception that stopped us after {time.monotonic()-t0} "
+                              f"seconds...")
+                self.stop()
+                totprocessed += self._drain_queue( batch )
+                batch = []
+                LOGGER.error( f"...processed {totprocessed} messages, raising an exception." )
                 raise

             else:
                 self._queue.task_done()
                 count += 1

-            if count == self.batch_maxn:
-                # hit the max number of results. process the batch
-                self.batch_callback(batch)
+            if ( (t1-t0) > self.batch_max_wait_between_messages ) or ( count >= self.batch_maxn ):
+                # hit the timeout or the max number of results. process the batch
+                t0 = t1
+                if len(batch) > 0:
+                    self.batch_callback(batch)
+                totprocessed += len(batch)
                 batch, count = [], 0

+            if ( ( ( max_runtime is not None ) and ( t1 - tstart >= max_runtime ) )
+                 or
+                 ( ( max_nmsgs is not None ) and ( totprocessed >= max_nmsgs ) )
+                ):
+                LOGGER.info( f"Stopping streaming after {t1-tstart:.1f} sec..." )
+                self.stop()
+                totprocessed += self._drain_queue( batch )
+                batch = []
+                LOGGER.info( f"...processed {totprocessed} messages, exiting." )
+                return totprocessed
+
+            # Listen for a die command, and send the heartbeat
+            if pipe is not None:
+                if ( datetime.datetime.now() - lastheartbeat ).total_seconds() > heartbeat:
+                    lastheartbeat = datetime.datetime.now()
+                    if pipe.poll():
+                        msg = pipe.recv()
+                        if ( 'command' in msg ) and ( msg['command'] == 'die' ):
+                            LOGGER.info( f"Received die command after {t1-tstart:.1f} sec..." )
+                            self.stop()
+                            totprocessed += self._drain_queue( batch )
+                            batch = []
+                            LOGGER.info( f"...processed {totprocessed} messages, exiting." )
+                            return totprocessed
+                    pipe.send( { "message": "ok", "nconsumed": totprocessed,
+                                 "runtime": lastheartbeat - firstheartbeat } )
+
     def stop(self) -> None:
         """Attempt to shutdown the streaming pull and exit the background threads gracefully."""
         LOGGER.info("closing the stream")
